File: src/com/scheduler/finance/module/BullishMomentumDetector.java
package com.scheduler.finance.module;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import com.scheduler.finance.vo.BullishMomentumSignal;
import com.scheduler.finance.vo.StockDataVo;

/**
 * 이동평균선 정배열 + 장대양봉/연속양봉 패턴 감지 모듈
 *
 * 패턴 설명:
 * - 완전 정배열: close > MA5 > MA20 > MA60 > MA120 > MA240
 * - 장대양봉: 실체(close-open)가 최근 평균 실체의 2배 이상
 * - 연속양봉: 3일 이상 연속 양봉(close > open)
 *
 * 이 패턴은 상승 추세 초입(Golden Cross) + 강한 매수 추진력을 나타낸다.
 */
public class BullishMomentumDetector {
    // WF-2-2: 차트 분석 모듈(모멘텀 신호)

    private static final int[] DEFAULT_PERIODS = {5, 20, 60, 120, 240};
    private static final int LOOKBACK_PERIOD = 20;  // 평균 실체 계산 기간

    /**
     * 완전 정배열 + 장대양봉/연속양봉 패턴 감지
     *
     * @param candles 캔들 데이터 리스트 (시간순 정렬)
     * @param stockCode 종목 코드
     * @param stockName 종목명 (선택)
     * @param consecutiveBullDays 연속양봉 일수 기준 (기본 3일)
     * @param longCandleRatio 장대양봉 배율 기준 (기본 2.0배)
     * @return 감지된 신호 리스트
     */
    public static List<BullishMomentumSignal> detectBullishMomentum(
            List<StockDataVo> candles,
            String stockCode,
            String stockName,
            int consecutiveBullDays,
            double longCandleRatio) {

        List<BullishMomentumSignal> signals = new ArrayList<>();

        if (candles == null || candles.isEmpty()) {
            return signals;
        }

        int size = candles.size();

        // MA 계산
        double[] ma5 = calculateSMA(candles, 5);
        double[] ma20 = calculateSMA(candles, 20);
        double[] ma60 = calculateSMA(candles, 60);
        double[] ma120 = calculateSMA(candles, 120);
        double[] ma240 = calculateSMA(candles, 240);

        // 완전 정배열 지점부터 시작
        int startIdx = 240;  // ma240을 위해 최소 240개 필요

        for (int i = startIdx; i < size; i++) {
            StockDataVo current = candles.get(i);
            double closePrice = current.getClose();

            // 1. 완전 정배열 체크: close > ma5 > ma20 > ma60 > ma120 > ma240
            if (!isCompleteAlignment(closePrice, ma5[i], ma20[i], ma60[i], ma120[i], ma240[i])) {
                continue;
            }

            // 2. 장대양봉 체크
            double avgBody = calculateAverageBody(candles, i);
            double currentBody = Math.abs(current.getClose() - current.getOpen());
            double bodyRatio = avgBody > 0 ? currentBody / avgBody : 0;

            if (bodyRatio >= longCandleRatio && current.getClose() > current.getOpen()) {
                BullishMomentumSignal signal = createSignal(
                        current, stockCode, stockName, "LONG_BULL_BODY",
                        ma5[i], ma20[i], ma60[i], ma120[i], ma240[i],
                        bodyRatio, 1
                );
                signals.add(signal);
            }

            // 3. 연속양봉 체크
            int consecutiveDays = countConsecutiveBullish(candles, i);
            if (consecutiveDays >= consecutiveBullDays) {
                BullishMomentumSignal signal = createSignal(
                        current, stockCode, stockName, "CONSECUTIVE_BULL",
                        ma5[i], ma20[i], ma60[i], ma120[i], ma240[i],
                        0, consecutiveDays
                );
                signals.add(signal);
            }
        }

        return signals;
    }

    /**
     * Chart List 형식(Highcharts 호환)에서 신호 감지
     *
     * @param chartList [time, open, high, low, close, volume] 구조의 리스트
     * @param stockCode 종목 코드
     * @param stockName 종목명
     * @param consecutiveBullDays 연속양봉 일수
     * @param longCandleRatio 장대양봉 배율
     * @return 감지된 신호 리스트
     */
    @SuppressWarnings("rawtypes")
    public static List<BullishMomentumSignal> detectBullishMomentumFromChartList(
            List<?> chartList,
            String stockCode,
            String stockName,
            int consecutiveBullDays,
            double longCandleRatio) {

        if (chartList == null || chartList.isEmpty()) {
            return new ArrayList<>();
        }

        List<StockDataVo> candles = new ArrayList<>();
        for (Object item : chartList) {
            if (!(item instanceof List)) {
                continue;
            }
            List row = (List) item;
            if (row.size() < 5) {
                continue;
            }

            try {
                long time = ((Number) row.get(0)).longValue();
                double open = ((Number) row.get(1)).doubleValue();
                double high = ((Number) row.get(2)).doubleValue();
                double low = ((Number) row.get(3)).doubleValue();
                double close = ((Number) row.get(4)).doubleValue();
                double volume = row.size() > 5 ? ((Number) row.get(5)).doubleValue() : 0;

                StockDataVo vo = new StockDataVo(time, open, high, low, close, volume);
                candles.add(vo);
            } catch (Exception e) {
                // Skip invalid row
                continue;
            }
        }

        return detectBullishMomentum(candles, stockCode, stockName, consecutiveBullDays, longCandleRatio);
    }

    /**
     * 완전 정배열 확인: close > ma5 > ma20 > ma60 > ma120 > ma240
     */
    private static boolean isCompleteAlignment(double close, double ma5, double ma20,
                                                double ma60, double ma120, double ma240) {
        if (Double.isNaN(ma5) || Double.isNaN(ma20) || Double.isNaN(ma60) ||
                Double.isNaN(ma120) || Double.isNaN(ma240)) {
            return false;
        }
        return close > ma5 && ma5 > ma20 && ma20 > ma60 && ma60 > ma120 && ma120 > ma240;
    }

    /**
     * 평균 실체 크기 계산 (최근 LOOKBACK_PERIOD일)
     */
    private static double calculateAverageBody(List<StockDataVo> candles, int endIdx) {
        int lookback = Math.min(LOOKBACK_PERIOD, endIdx + 1);
        if (lookback <= 0) {
            return 0;
        }

        double sum = 0;
        for (int i = endIdx - lookback + 1; i <= endIdx; i++) {
            if (i >= 0 && i < candles.size()) {
                StockDataVo candle = candles.get(i);
                sum += Math.abs(candle.getClose() - candle.getOpen());
            }
        }
        return sum / lookback;
    }

    /**
     * 현재 지점부터 거슬러 올라가며 연속 양봉 일수 카운트
     */
    private static int countConsecutiveBullish(List<StockDataVo> candles, int endIdx) {
        int count = 0;
        for (int i = endIdx; i >= 0; i--) {
            StockDataVo candle = candles.get(i);
            if (candle.getClose() > candle.getOpen()) {
                count++;
            } else {
                break;
            }
        }
        return count;
    }

    /**
     * 신호 객체 생성
     */
    private static BullishMomentumSignal createSignal(
            StockDataVo current,
            String stockCode,
            String stockName,
            String signalType,
            double ma5,
            double ma20,
            double ma60,
            double ma120,
            double ma240,
            double bodyRatio,
            int consecutiveDays) {

        BullishMomentumSignal signal = new BullishMomentumSignal();
        signal.setSignalTime(current.getDate());
        signal.setStockCode(stockCode);
        signal.setStockName(stockName);
        signal.setClosePrice(current.getClose());
        signal.setSignalType(signalType);
        signal.setOpen(current.getOpen());
        signal.setHigh(current.getHigh());
        signal.setLow(current.getLow());
        signal.setVolume(current.getVolume());
        signal.setMa5(ma5);
        signal.setMa20(ma20);
        signal.setMa60(ma60);
        signal.setMa120(ma120);
        signal.setMa240(ma240);
        signal.setBodyRatio(bodyRatio);
        signal.setConsecutiveDays(consecutiveDays);

        return signal;
    }

    /**
     * 단순 이동평균(SMA) 계산
     */
    private static double[] calculateSMA(List<StockDataVo> candles, int period) {
        int n = candles.size();
        double[] sma = new double[n];
        Arrays.fill(sma, Double.NaN);

        if (period <= 0 || n < period) {
            return sma;
        }

        double sum = 0.0;
        for (int i = 0; i < n; i++) {
            sum += candles.get(i).getClose();
            if (i >= period) {
                sum -= candles.get(i - period).getClose();
            }
            if (i >= period - 1) {
                sma[i] = sum / period;
            }
        }
        return sma;
    }

    /**
     * 여러 종목 차트 데이터에서 일괄 신호 감지
     */
    public static List<BullishMomentumSignal> detectBullishMomentumBatch(
            Map<String, List<StockDataVo>> candlesByStock,
            Map<String, String> stockNames,
            int consecutiveBullDays,
            double longCandleRatio) {

        List<BullishMomentumSignal> allSignals = new ArrayList<>();

        for (Map.Entry<String, List<StockDataVo>> entry : candlesByStock.entrySet()) {
            String stockCode = entry.getKey();
            List<StockDataVo> candles = entry.getValue();
            String stockName = stockNames != null ? stockNames.get(stockCode) : null;

            List<BullishMomentumSignal> signals = detectBullishMomentum(
                    candles, stockCode, stockName, consecutiveBullDays, longCandleRatio
            );
            allSignals.addAll(signals);
        }

        return allSignals;
    }
}

File: webapp/appone/jsp/finance/kis/kisFinance/js/bullishMomentumDetector.js
/**
 * 이동평균선 정배열 + 장대양봉/연속양봉 패턴 감지 클라이언트 라이브러리
 *
 * 사용 예:
 * const detector = new BullishMomentumDetector({
 *     consecutiveBullDays: 3,
 *     longCandleRatio: 2.0
 * });
 *
 * detector.detectFromChart(chartData)
 *     .then(signals => {
 *         console.log('감지된 신호:', signals);
 *         detector.displaySignalsOnChart(chart, signals);
 *     });
 */
class BullishMomentumDetector {
    constructor(options) {
        this.options = {
            consecutiveBullDays: options?.consecutiveBullDays || 3,
            longCandleRatio: options?.longCandleRatio || 2.0,
            apiEndpoint: options?.apiEndpoint || '/scheduler/finance/detectBullishMomentum.do'
        };
    }

    /**
     * API를 통해 신호 감지 (서버 처리)
     */
    detectFromChart(chartData, stockCode, stockName) {
        const params = new URLSearchParams();
        params.append('stockCode', stockCode || '');
        params.append('stockName', stockName || '');
        params.append('chartData', JSON.stringify({ data: chartData }));
        params.append('consecutiveBullDays', this.options.consecutiveBullDays);
        params.append('longCandleRatio', this.options.longCandleRatio);

        return fetch(this.options.apiEndpoint, {
            method: 'POST',
            body: params,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            }
        })
        .then(response => response.json())
        .then(result => {
            if (result.success) {
                return result.signals || [];
            } else {
                console.error('신호 감지 실패:', result.message);
                return [];
            }
        })
        .catch(error => {
            console.error('API 호출 오류:', error);
            return [];
        });
    }

    /**
     * 클라이언트 처리: StockDataVo 배열에서 신호 감지
     */
    detectLocalFromStockData(candles, stockCode, stockName) {
        if (!candles || candles.length < 240) {
            console.warn(`충분한 캔들 데이터 필요: ${candles?.length || 0} / 240`);
            return [];
        }

        const ma5 = this._calculateSMA(candles, 5);
        const ma20 = this._calculateSMA(candles, 20);
        const ma60 = this._calculateSMA(candles, 60);
        const ma120 = this._calculateSMA(candles, 120);
        const ma240 = this._calculateSMA(candles, 240);

        const signals = [];
        const startIdx = 240;

        for (let i = startIdx; i < candles.length; i++) {
            const current = candles[i];
            const close = current.close;

            // 완전 정배열 확인
            if (!this._isCompleteAlignment(close, ma5[i], ma20[i], ma60[i], ma120[i], ma240[i])) {
                continue;
            }

            // 장대양봉 확인
            const avgBody = this._calculateAverageBody(candles, i);
            const currentBody = Math.abs(current.close - current.open);
            const bodyRatio = avgBody > 0 ? currentBody / avgBody : 0;

            if (bodyRatio >= this.options.longCandleRatio && current.close > current.open) {
                signals.push({
                    signalTime: current.date,
                    stockCode: stockCode,
                    stockName: stockName,
                    closePrice: current.close,
                    signalType: 'LONG_BULL_BODY',
                    bodyRatio: bodyRatio,
                    consecutiveDays: 1,
                    ma5: ma5[i],
                    ma20: ma20[i],
                    ma60: ma60[i],
                    ma120: ma120[i],
                    ma240: ma240[i],
                    open: current.open,
                    high: current.high,
                    low: current.low,
                    volume: current.volume
                });
            }

            // 연속양봉 확인
            const consecutiveDays = this._countConsecutiveBullish(candles, i);
            if (consecutiveDays >= this.options.consecutiveBullDays) {
                signals.push({
                    signalTime: current.date,
                    stockCode: stockCode,
                    stockName: stockName,
                    closePrice: current.close,
                    signalType: 'CONSECUTIVE_BULL',
                    bodyRatio: 0,
                    consecutiveDays: consecutiveDays,
                    ma5: ma5[i],
                    ma20: ma20[i],
                    ma60: ma60[i],
                    ma120: ma120[i],
                    ma240: ma240[i],
                    open: current.open,
                    high: current.high,
                    low: current.low,
                    volume: current.volume
                });
            }
        }

        return signals;
    }

    /**
     * Highcharts 차트에 신호 마킹
     */
    displaySignalsOnChart(chart, signals) {
        if (!chart || !signals || signals.length === 0) {
            return;
        }

        signals.forEach(signal => {
            const color = signal.signalType === 'LONG_BULL_BODY' ? '#FF6B6B' : '#4ECDC4';
            const text = signal.signalType === 'LONG_BULL_BODY'
                ? `장대양봉 ${signal.bodyRatio.toFixed(1)}배`
                : `연속양봉 ${signal.consecutiveDays}일`;

            // Highcharts에 label 추가
            chart.renderer.label(text, 0, 0)
                .attr({
                    fill: color,
                    padding: 5,
                    r: 3,
                    zIndex: 6
                })
                .css({
                    color: 'white',
                    fontSize: '11px',
                    fontWeight: 'bold'
                })
                .add();
        });

        console.log(`차트에 ${signals.length}개 신호 마킹`);
    }

    /**
     * 신호를 HTML 테이블로 렌더링
     */
    renderSignalsTable(signals, containerId) {
        const container = document.getElementById(containerId);
        if (!container) {
            console.error(`Container not found: ${containerId}`);
            return;
        }

        if (!signals || signals.length === 0) {
            container.innerHTML = '<p>감지된 신호가 없습니다.</p>';
            return;
        }

        let html = `
            <table class="table table-sm table-striped">
                <thead class="table-dark">
                    <tr>
                        <th>시간</th>
                        <th>종목</th>
                        <th>종가</th>
                        <th>신호 유형</th>
                        <th>상세</th>
                        <th>MA 값</th>
                    </tr>
                </thead>
                <tbody>
        `;

        signals.forEach(signal => {
            const date = new Date(signal.signalTime).toLocaleDateString('ko-KR');
            const typeLabel = signal.signalType === 'LONG_BULL_BODY'
                ? `<span class="badge bg-danger">장대양봉</span>`
                : `<span class="badge bg-info">연속양봉</span>`;

            const details = signal.signalType === 'LONG_BULL_BODY'
                ? `${signal.bodyRatio.toFixed(2)}배`
                : `${signal.consecutiveDays}일`;

            const maInfo = `
                5: ${signal.ma5?.toFixed(0)}<br/>
                20: ${signal.ma20?.toFixed(0)}<br/>
                60: ${signal.ma60?.toFixed(0)}<br/>
                120: ${signal.ma120?.toFixed(0)}<br/>
                240: ${signal.ma240?.toFixed(0)}
            `;

            html += `
                <tr>
                    <td>${date}</td>
                    <td>${signal.stockCode} ${signal.stockName || ''}</td>
                    <td>${signal.closePrice?.toFixed(0)}</td>
                    <td>${typeLabel}</td>
                    <td>${details}</td>
                    <td style="font-size: 0.85em;">${maInfo}</td>
                </tr>
            `;
        });

        html += `
                </tbody>
            </table>
        `;

        container.innerHTML = html;
    }

    // ======= 비공개 헬퍼 메서드 =======

    _calculateSMA(candles, period) {
        const sma = new Array(candles.length).fill(NaN);
        if (period <= 0 || candles.length < period) {
            return sma;
        }

        let sum = 0;
        for (let i = 0; i < candles.length; i++) {
            sum += candles[i].close;
            if (i >= period) {
                sum -= candles[i - period].close;
            }
            if (i >= period - 1) {
                sma[i] = sum / period;
            }
        }
        return sma;
    }

    _isCompleteAlignment(close, ma5, ma20, ma60, ma120, ma240) {
        return !isNaN(ma5) && !isNaN(ma20) && !isNaN(ma60) && !isNaN(ma120) && !isNaN(ma240)
            && close > ma5 && ma5 > ma20 && ma20 > ma60 && ma60 > ma120 && ma120 > ma240;
    }

    _calculateAverageBody(candles, endIdx) {
        const lookback = Math.min(20, endIdx + 1);
        if (lookback <= 0) return 0;

        let sum = 0;
        for (let i = endIdx - lookback + 1; i <= endIdx; i++) {
            if (i >= 0 && i < candles.length) {
                sum += Math.abs(candles[i].close - candles[i].open);
            }
        }
        return sum / lookback;
    }

    _countConsecutiveBullish(candles, endIdx) {
        let count = 0;
        for (let i = endIdx; i >= 0; i--) {
            if (candles[i].close > candles[i].open) {
                count++;
            } else {
                break;
            }
        }
        return count;
    }
}

// 글로벌 인스턴스 내보내기
if (typeof window !== 'undefined') {
    window.BullishMomentumDetector = BullishMomentumDetector;
}
